File: Semester2/2/task4/task4/ArrayStack.cs
using System;

namespace Task4
{
    /// <summary>
    /// Stack based on array
    /// </summary>
    class ArrayStack<T> : InterfaceStack<T> 
    {
        public ArrayStack(int maxSize)
        {
            this.currentSize = 0;
            this.maxSize = maxSize;
            this.array = new T[maxSize];
        }
        
        private int maxSize;
        private int currentSize;
        private T[] array;

        public void Push(T value)
        {
            array[currentSize] = value;
            currentSize++;
        }

        public void PrintStack()
        {
            for (int i = 0; i < currentSize; i++)
                Console.Write(array[i] + " ");
            Console.WriteLine();
            return;
        }

        public T Pop()
        {
            if (IsEmpty())
                return default(T);
            T auxilary = array[currentSize - 1];
            currentSize--;
            return auxilary;
        }

        public T Top()
        {
            if (IsEmpty())
                return default(T);
            return array[currentSize - 1];
        }

        public bool IsEmpty()
        {
            return currentSize == 0;
        }
    }
}

File: qt/Calc/arrayStack.cpp
#include "arrayStack.h"
#include "myException.h"

ArrayStack::ArrayStack(int maxSize) :
    mCurrent(0)
  , mArray(new double[maxSize])
{
}

void ArrayStack::push(double value)
{
    mArray[mCurrent] = value;
    mCurrent++;
}

double ArrayStack::pop()
{
    if (size() == 0)
    {
        throw MyException("Stack is empty");
    }

    mCurrent--;
    return mArray[mCurrent + 1];
}

double ArrayStack::top() const
{
    return mArray[mCurrent];
}

int ArrayStack::size() const
{
    return mCurrent;
}

ArrayStack::~ArrayStack()
{
    mCurrent = 0;
    delete []mArray;
}
